duckduckgo/BrowserServicesKit
# Automatically save generated passwords

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1201969773101195/f
iOS PR:  https://github.com/duckduckgo/iOS/pull/1109
macOS PR: https://github.com/more-duckduckgo-org/macos-browser/pull/500

**Description**:

This PR updates BSK to automatically save generated credentials when the user selects that option. Note that this PR does not require any changes on either of the client codebases.

If the user doesn't have a username entered, they will get prompted immediately. If they _have_ already entered a username, the credential gets saved without prompting them. See [this task](https://app.asana.com/0/1175293949586521/1202076731691825/f) for context.

**Steps to test this PR**:
1. Visit a site that has a sign-up form; I've been using [Hacker News](http://news.ycombinator.com/login) since it's a super simple page
1. Click into the password field, and click the generated password prompt, and ensure you have no username
1. You should get a prompt after clicking the generated password field; go ahead and save that
1. Go back to the sign up form, type in a username, and hit enter (it's fine to use an existing username to prevent creating an actual account, the app just needs the form submission event to trigger)
1. Check the Autofill panel and verify that the previous login has been updated to include the username



**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
